### ISSUE TITLE
Cached data is not available on the QueryResult when there is a network error.

### ISSUE THREAD

**Issue by jifalops** [zino-app/graphql-flutter#718]
**Describe the bug**
Cached data is not available on the QueryResult when there is a network error.

**To Reproduce**
1. Set-up
```dart
GraphQLProvider(
      client: ValueNotifier<GraphQLClient>(
        GraphQLClient(
          cache: OptimisticCache(dataIdFromObject: _idFromObject),
          link: myLink,
        ),
      ),
      child: CacheProvider(
        child: MaterialApp(
  ...
```
2. Run a query and get the results.
```dart
class MyQuery extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Query(
      options: QueryOptions(
        // query is of a type generated by the `artemis` package
        documentNode: query.document,
      ),
      builder: (
        QueryResult result, {
        Future<QueryResult> Function() refetch,
        FetchMore fetchMore,
      }) {
        if (result.hasException) {
          return Text('${result.exception}');
        } else if (result.loading) {
          return Center(child: CircularProgressIndicator());
        }

        final items = query.parse(result.data);

        return Column(
          children: [
            ...items.map((item) => ListTile(
                  title: Text(item.name),
                  subtitle: Text('${item.typeName} ${item.id}'),
                ))
          ],
        );
      },
    );
  }
}
```
3. Enable airplane mode on the device
4. Press home and open another app (so the cache write is triggered)
5. Restart the app (hot restart). The result is now the network error and no cached data is available.

**Expected behavior**
Enabling airplane mode and hot-restarting the app should allow it to return the cached results.

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04

**Smartphone (please complete the following information):**
 - Device: Android emulator API 29

**Additional context**
Flutter beta 1.21.0-9.2.pre

**Comment by micimize** [comment]
Can you try this with `4.0.0-alpha.5`?

**Comment by micimize** [comment]
This is outdated / likely solved by the cache rework `4.0.0-beta.1` (#648). Also this may just be because of the hot restart.

Please make a reproduction that doesn't involve live reload for testing

**Comment by micimize** [comment]
looked at this again and added a mechanism for carrying forward errors in #754

**Comment by jifalops** [comment]
i actually solved it with an HTTP link that caches queries. I think that is what's right for my use case.

**Comment by micimize** [comment]
Closed by #754
